include/unifex/linux/io_epoll_context.hpp

```diff
@@ -0,0 +1,520 @@
+
+#include <sys/epoll.h>
+#include <sys/timerfd.h>
+#include <sys/uio.h>
+
+namespace unifex {
+namespace linuxos {
+
+class io_epoll_context {
+ public:
+  class schedule_sender;
+  class schedule_at_sender;
+  template <typename Duration>
+  class schedule_after_sender;
+  class read_sender;
+  class write_sender;
+  class async_read_only_file;
+  class async_read_write_file;
+  class async_write_only_file;
+  class scheduler;
+
+  io_epoll_context();
+
+  ~io_epoll_context();
+
+  template <typename StopToken>
+  void run(StopToken stopToken);
+
+  scheduler get_scheduler() noexcept;
+
+ private:
+  struct operation_base {
+    operation_base() noexcept {}
+    operation_base* next_;
+    void (*execute_)(operation_base*) noexcept;
+  };
+
+  struct completion_base : operation_base {
+    int result_;
+  };
+
+  struct stop_operation : operation_base {
+    stop_operation() noexcept {
+      this->execute_ = [](operation_base * op) noexcept {
+        static_cast<stop_operation*>(op)->shouldStop_ = true;
+      };
+    }
+    bool shouldStop_ = false;
+  };
+
+  using time_point = linuxos::monotonic_clock::time_point;
+
+  struct schedule_at_operation : operation_base {
+    explicit schedule_at_operation(
+        io_epoll_context& context,
+        const time_point& dueTime,
+        bool canBeCancelled) noexcept
+        : context_(context),
+          dueTime_(dueTime),
+          canBeCancelled_(canBeCancelled) {}
+
+    schedule_at_operation* timerNext_;
+    schedule_at_operation* timerPrev_;
+    io_epoll_context& context_;
+    time_point dueTime_;
+    bool canBeCancelled_;
+
+    static constexpr std::uint32_t timer_elapsed_flag = 1;
+    static constexpr std::uint32_t cancel_pending_flag = 2;
+    std::atomic<std::uint32_t> state_ = 0;
+  };
+
+  using operation_queue =
+      intrusive_queue<operation_base, &operation_base::next_>;
+
+  using timer_heap = intrusive_heap<
+      schedule_at_operation,
+      &schedule_at_operation::timerNext_,
+      &schedule_at_operation::timerPrev_,
+      time_point,
+      &schedule_at_operation::dueTime_>;
+
+  bool is_running_on_io_thread() const noexcept;
+  void run_impl(const bool& shouldStop);
+
+  void schedule_impl(operation_base* op);
+  void schedule_local(operation_base* op) noexcept;
+  void schedule_local(operation_queue ops) noexcept;
+  void schedule_remote(operation_base* op) noexcept;
+
+  // Insert the timer operation into the queue of timers.
+  // Must be called from the I/O thread.
+  void schedule_at_impl(schedule_at_operation* op) noexcept;
+
+  // Execute all ready-to-run items on the local queue.
+  // Will not run other items that were enqueued during the execution of the
+  // items that were already enqueued.
+  // This bounds the amount of work to a finite amount.
+  void execute_pending_local() noexcept;
+
+  // Check if any completion queue items are available and if so add them
+  // to the local queue.
+  void acquire_completion_queue_items() noexcept;
```

Comment:
I think this method should probably be modified to incorporate the call to `epoll_wait()` and have it allocate the buffer of `epoll_event` objects on the stack instead of as data-members.

Also, consider renaming this method to something more epoll-specific (the "completion queue" terminology comes from io_uring).